.dev/docker/env/Dockerfile

```diff
@@ -1,33 +1,41 @@
-# For valid combinations, check the following repo:
-# https://gitlab.com/nvidia/container-images/cuda/tree/master/dist
-# To enable cuda, use "--build-arg USE_CUDA=true" during image build process
-ARG USE_CUDA
-ARG CUDA_VERSION="9.2"
-ARG UBUNTU_DISTRO="16.04"
-ARG BASE_CUDA_IMAGE=${USE_CUDA:+"nvidia/cuda:${CUDA_VERSION}-devel-ubuntu${UBUNTU_DISTRO}"}
-ARG BASE_IMAGE=${BASE_CUDA_IMAGE:-"ubuntu:${UBUNTU_DISTRO}"}
+ARG UBUNTU_VERSION=20.04
 
-FROM ${BASE_IMAGE}
+FROM "ubuntu:${UBUNTU_VERSION}"
 
+# Eigen patch (https://eigen.tuxfamily.org/bz/show_bug.cgi?id=1462) to fix issue metioned
+# in https://github.com/PointCloudLibrary/pcl/issues/3729 is available in Eigen 3.3.7.
+# Not needed from 20.04 since it is the default version from apt
+ARG EIGEN_MINIMUM_VERSION=3.3.7
+
+# See https://www.optonic.com/support/download/ensenso-sdk/archiv/ for available versions
+ARG ENSENSOSDK_VERSION=3.2.489
```

Comment:
The default value does not seem to be used correctly https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20654&view=logs&j=02e31b63-97df-57da-5c5b-409d1f33cda8&t=3198efe3-d7f9-5c6d-757b-14de2e7a3e6f

Reply:
The issue is, that we pass `ENSENSOSDK_VERSION=""` (an empty value) via `env.yml`. Therefore the default value is not used. I think we would have the same issue with `librealsense`, when would have passed a value.

Should an empyt value means: Don't install it or use default?

Reply:
Either seems fine. If empty value=don't install, then we have to specify `ENSENSOSDK_VERSION: 3.2.489` for every Ubuntu version in env.yml, because we do want it installed, right?

Reply:
Yep.

Played a bit with the Dockerfile now.

Sadly this is not working:
```Dockerfile
ARG ENSENSOSDK_VERSION=${ENSENSOSDK_VERSION:+"3.2.489"}
```
As the [parameter expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) doesn't have any check for an empty value (and not just null or unset).

So we would need to do sth. like (see [here](https://gist.github.com/JoeyBurzynski/f185285602b3c9f0fe708e32ad3ea06f)):
```sh
if [ -z "${ENSENSOSDK_VERSION}" ]; then
```
Therefore we loose the default value at beginning of the Dockerfile or need to define it twice.

Sadly I'm not Azure pipeline expert, but is it possible to do sth. like this?

Instead of just:
```yml
  - task: Docker@2
    displayName: "Build docker image"
    inputs:
      command: build
      arguments: |
        --no-cache
        --build-arg UBUNTU_VERSION=$(UBUNTU_VERSION)
        --build-arg ENSENSOSDK_VERSION=$(ENSENSOSDK_VERSION)
        --build-arg VTK_VERSION=$(VTK_VERSION)
        --build-arg USE_LATEST_CMAKE=$(USE_LATEST_CMAKE)
        -t $(dockerHubID)/env:$(TAG)
      dockerfile: '$(Build.SourcesDirectory)/.dev/docker/env/Dockerfile'
      tags: "$(TAG)"
```

```yml
  - task: Docker@2
    displayName: "Build docker image"
   script:
        dockerBuildArgs=""
        if [ -z "${UBUNTU_VERSION}" ]; then
          dockerBuildArgs="$dockerBuildArgs --build-arg UBUNTU_VERSION=$(UBUNTU_VERSION)"
        fi
        if [ -z "${ENSENSOSDK_VERSION}" ]; then
          dockerBuildArgs="$dockerBuildArgs --build-arg ENSENSOSDK_VERSION=$(ENSENSOSDK_VERSION)"
        fi
        if [ -z "${VTK_VERSION}" ]; then
          dockerBuildArgs="$dockerBuildArgs --build-arg VTK_VERSION=$(VTK_VERSION)"
        fi
        if [ -z "${USE_LATEST_CMAKE}" ]; then
          dockerBuildArgs="$dockerBuildArgs --build-arg USE_LATEST_CMAKE=$(USE_LATEST_CMAKE)"
        fi
 - script: |
    inputs:
      command: build
      arguments: |
        --no-cache
        --build-arg UBUNTU_VERSION=$(UBUNTU_VERSION)
        --build-arg ENSENSOSDK_VERSION=$(ENSENSOSDK_VERSION)
        --build-arg VTK_VERSION=$(VTK_VERSION)
        --build-arg USE_LATEST_CMAKE=$(USE_LATEST_CMAKE)
        -t $(dockerHubID)/env:$(TAG)
      dockerfile: '$(Build.SourcesDirectory)/.dev/docker/env/Dockerfile'
      tags: "$(TAG)"
```
?
So prepending a script within the job

Reply:
No idea if that works. Test it if you like, but I would also be ok with explicitly specifying the ensenso sdk version for every Ubuntu version (advantage: makes it easier to track which version is installed in each docker image)

Reply:
It now says that "All checks have passed", but the Docker pipeline is not listed there and I suspect that there is some reporting error. Could you try to push something again?

Reply:
Indeed, this is a reporting issue. It is [still not working](https://stackoverflow.com/questions/70609660/pass-argument-only-to-docker-task-if-value-is-not-empty) :-(

Reply:
Currently the most issue is: The values from the matrix strategy are paramaters - and if they are not set Azure doesn't replace `$(ENSENSOSDK_VERSION)`. It looks like the are doing search & replace when parsing :-/

Reply:
I used now a separate bash script to build the arguments. Maybe we can improve it later when we get a suggestion from my question on StackOverflow.

It seems to working . except 20.10, because of `The repository 'http://archive.ubuntu.com/ubuntu groovy Release' does not have a Release file.` 😆

Reply:
Ok, they removed the update source of 20.10. So I switched everything to 21.10.